### PR TITLE
Feature/enumerables branch

### PR DESCRIPTION
**Pull request by Paul-Clue**
## What we did;
- We made clones of functions in the Enumerable class.
- The naming convention of our functions simply adds my_ to the front of the name of the Enumerable function it is meant to mimic.
- We did tests to make sure our functions worked.

### CODE REVIEW

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️

### Hello @Paul-Clue and @franklinben23, your progress is good

However, there are still some changes that are necessary for this project to be approved.
Let us address them and request another review.

- [x] `#my_each` should returns an Enumerator if no block is given
- [x] `#my_each` when a block is given when self is an array should return the array itself after calling the given block once for each element in self
- [x] `#my_each` when a block is given when self is a range returns the range itself after calling the given block once for each element in self
- [x] `#my_each_with_index` returns an enumerator if no block is given
- [x] `#my_select` returns an enumerator if no block is given

- [x] `#my_select` returns an enumerator if no block is given
`[1,2,3,4].my_select.class==Enumerator`

- [x] `#my_all?` when no block or argument is given returns true when none of the collection members are false or nil
 `[1, true, "hi", []].all? ==  [1, true, "hi", []].my_all?`

- [x] `#my_all?` when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern
`[5,5,[5]].all?(5) == [5,5,[5]].my_all?(5)`

- [x] `#my_none?` when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
`['dog','car'].none?(5) == ['dog','car'].my_none?(5)`

- [x] `#my_none?` when a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern
`%w[dog door rod blade].none?(5) == %w[dog door rod blade].my_none?(5)`

- [x] `#my_map` should return an Enumerator if no block is given
- [ ] `#my_map` executes only the proc when both a block and a proc are given and should return a new array with the results of running the proc once for every element in enum.

- [x] `#my_inject`` raises a "LocalJumpError" when no block or argument is given
     `Failure/Error: expect { array.my_inject }.to raise_error(LocalJumpError)
       expected LocalJumpError but nothing was raised`

- [ ] `#my_inject` searches for the longest word in an array of strings
`['to','four','tri'].my_inject(&proc{|a,b| [a,b].max_by{|m| m.length}}) == ['to','four','tri'].inject(&proc{|a,b| [a,b].max_by{|m| m.length}})`

- [x] `#my_inject` does not mutate the original array 
```
array = [1,2,3,4]
array_clone = array.clone
array.my_inject{|total, num| total+num}
array == array_clone
```

- [x]   `#my_inject` when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block
`array.my_inject(&operation).to eq array.inject(&operation)`

- [x] `#my_inject` when a symbol is specified without an initail value combines each element of the collection by applying the symbol as a named method
`[1,2,3].inject(:+) == [1,2,3].my_inject(:+)`

 
- [x] `#my_inject` when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method:range
`(1..9).inject(:+) == (1..9).my_inject(:+)`

- [x] `#my_inject` when a symbol is specified with an initail value combines each element of the collection by applying the symbol as a named method
`[1,2,3].inject(4, :*) == [1,2,3].my_inject(4, :*)`

- [x] `#my_inject` when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method::range
`(1..3).inject(4, :*) == (1..3).my_inject(4, :*)`

 
- [x] #multiply_els accepts an array as an argument and multiplies all the elements of the array together using #my_inject
`multiply_els([1,2,3]) == 6`

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla), if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

File: enumerables.rb [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+  end
+
+  def my_each_with_index
+    return to_enum(:my_each_with_index) unless block_given?
+
+    h_array = to_a
+    0.upto(length - 1) do |v|
+      yield(h_array[v], v)
+    end
+  end
```
Comment (lekegitrepo) [inline enumerables.rb:32]: - **_`#my_each_with_index` when a block is given when self is an array returns the array itself after calling the given block once for each element in self_**
- **_`#my_each_with_index`, when a block is given when self is a range, returns the range itself after calling the given block once for each element in self_**
- **_`#my_each_with_index`, when a block is given when self is a range, behaves like `#each_with_index` calls the given block once for each element in self_**

File: enumerables.rb [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+    return to_enum(:double) unless block_given?
+
+    h_array = to_a
+    map_arr = []
+
+    h_array.my_each do |v|
+      map_arr.push(yield(v))
+    end
+    map_arr
+  end
```
Comment (lekegitrepo) [inline enumerables.rb:126]: - **_`#my_map` should execute only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum._**

File: enumerables.rb [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+      sum = block.yield(sum, h_range[index]) unless num.positive?
+      sum = block.yield(sum, h_range[index - 1]) unless num.positive? == false
+      index += 1
+      break unless (index == h_range.length) == false
+    end
+    return sum unless num.positive?
+
+    sum
+  end
+end
```
Comment (lekegitrepo) [inline enumerables.rb:169]: - **_#my_inject raises a "LocalJumpError" when no block or argument is given_**
- **_This example mentioned in the last review still failing:  `#my_inject` searches for the longest word in an array of strings
`['to','four','tri'].my_inject(&proc{|a,b| [a,b].max_by{|m| m.length}}) == ['to','four','tri'].inject(&proc{|a,b| [a,b].max_by{|m| m.length}})`_**
- **_`#my_inject`, when a block is given without an initial value, combines all elements of enum by applying a binary operation, specified by a block_**
- **_`#my_inject`, when a symbol is specified with an initial value, combines each element of the collection by applying the symbol as a named method_**
- **_`#my_inject`, when a symbol is specified with an initial value, combines each element of the collection by applying the symbol as a named method::range_**

**Review by lekegitrepo — CHANGES_REQUESTED** [review]
#### Status: Changes Required! :recycle:
#### Hi @Paul-Clue and @franklinben23 :clap: :clap:
#### Good job so far :+1: However, some of the methods are still failing. Check my comments below :point_down:

#### _Reviewed by Temitayo Adeleke_

File: enumerables.rb (hunk 1 above)
Comment (Paul-Clue) [reply enumerables.rb:32]: Hey, Temitayo. We just checked the my_each_with_index function and it does not return the self array. But the original function each_with_index does not return the self array either. Please check this.

File: enumerables.rb (hunk 2 above)
Comment (Paul-Clue) [reply enumerables.rb:126]: We checked this also and the original #map function does not prioritize a proc over a block. The original #map function returns an error if you pass it a proc and a block at the same time. And so does our #my_map function.

**Review by BrianSammit — CHANGES_REQUESTED** [review]
CHANGE REQUIRED :recycle: 
Great job team :+1: 
however, there are some changes that you have to work on. 
-----------------------------

- [ ] #my_each_with_index when a block is given when self is an array returns the array itself.
- [ ] #my_each_with_index when a block is given when self is a range returns the range itself after calling the given block once for each element in self
- [ ] #my_each_with_index when a block is given when self is a range behaves like #each_with_index calls the given block once for each element in self
- [ ] #my_all? when no block or argument is given returns false when one of the collection members are false or nil
- [ ] #my_all? when a class is passed as an argument returns true if all of the collection is a member of such class::Integer
- [ ] #my_all? when a class is passed as an argument returns true if all of the collection is a member of such class::Numeric
- [ ] #my_all? when a Regex is passed as an argument returns true if all of the collection matches the Regex
- [ ] #my_any? returns true if the block ever returns a value other than false or nil
- [ ] #my_any? returns false if the block ever returns a value false or nil
- [ ] #my_any? returns false if the block ever returns a value false or nil::range
- [ ] #my_any? when no block or argument is given returns true if at least one of the collection is not false or nil
- [ ] #my_any? when no block or argument is given returns false if at least one of the collection is not true
- [ ] #my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class::Numeric

------------------------------------------------------------------------------
Make the changes and request for another review :clap: 
I know this project is hard, you are doing excellent work keep doing it :muscle: 

HAPPY CODING 

MB5W4D

**Comment by Paul-Clue** [comment]
We ran tests on all of the functions that you mention here. all of our tests work and give the same output as the original functions that they are meant to mimic. Here is a link to a loom video showing one example: 
https://www.loom.com/share/ee3f8ee8856449169ee2d0840df43560

**Review by NtwaliHeritier — APPROVED** [review]
**STATUS: APPROVED ✔**
Hello team,
Great work implementing the changes. Feel free to merge and move on.
Congrats👌

**Pull request merged**